runtime/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaSessionFactory.java

```diff
@@ -3515,106 +3802,1263 @@ private void doKafkaData(
+            final long acknowledge = reset.acknowledge();
+            final long traceId = reset.traceId();
+
+            assert acknowledge <= sequence;
+
+            delegate.doMqttReset(traceId, EMPTY_OCTETS);
+        }
+
+        private void doKafkaData(
+            long traceId,
+            long authorization,
+            long budgetId,
+            int flags,
+            Flyweight extension)
+        {
+            doData(kafka, originId, routedId, initialId, initialSeq, initialAck, initialMax,
+                traceId, authorization, budgetId, flags, 0, EMPTY_OCTETS, extension);
+
+            assert initialSeq <= initialAck + initialMax;
+        }
+
+        private void doKafkaReset(
+            long traceId)
+        {
+            if (!MqttKafkaState.replyClosed(state))
+            {
+                state = MqttKafkaState.closeReply(state);
+
+                doReset(kafka, originId, routedId, replyId, replySeq, replyAck, replyMax, traceId, EMPTY_OCTETS);
+            }
+        }
+
+        private void doKafkaWindow(
+            long traceId,
+            long authorization,
+            long budgetId,
+            int padding,
+            int capabilities)
+        {
+            replyAck = delegate.replyAck;
+            replyMax = delegate.replyMax;
+            replyPad = delegate.replyPad;
+
+            doWindow(kafka, originId, routedId, replyId, replySeq, replyAck, replyMax,
+                traceId, authorization, budgetId, padding, replyPad, capabilities);
+        }
+    }
+
+    private PublishOffsetMetadata stringToOffsetMetadata(
+        String16FW metadata)
+    {
+        final IntArrayList packetIds = new IntArrayList();
+        UnsafeBuffer buffer = new UnsafeBuffer(BitUtil.fromHex(metadata.asString()));
+        final MqttPublishOffsetMetadataFW offsetMetadata = mqttOffsetMetadataRO.wrap(buffer, 0, buffer.capacity());
+        if (offsetMetadata.packetIds() != null)
+        {
+            offsetMetadata.packetIds().forEachRemaining((IntConsumer) packetIds::add);
+        }
+        return new PublishOffsetMetadata(offsetMetadata.producerId(), offsetMetadata.producerEpoch(), packetIds);
+    }
+
+    private String16FW offsetMetadataToString(
+        PublishOffsetMetadata metadata)
+    {
+        mqttOffsetMetadataRW.wrap(offsetBuffer, 0, offsetBuffer.capacity());
+        mqttOffsetMetadataRW.version(OFFSET_METADATA_VERSION);
+        mqttOffsetMetadataRW.producerId(metadata.producerId);
+        mqttOffsetMetadataRW.producerEpoch(metadata.producerEpoch);
+
+        if (metadata.packetIds != null)
+        {
+            metadata.packetIds.forEach(p -> mqttOffsetMetadataRW.appendPacketIds(p.shortValue()));
+        }
+        final MqttPublishOffsetMetadataFW offsetMetadata = mqttOffsetMetadataRW.build();
+        return new String16FW(BitUtil.toHex(offsetMetadata.buffer().byteArray(),
+            offsetMetadata.offset(), offsetMetadata.limit()));
+    }
```

Comment:
There is a lot of allocation going on here. Let's see if we can reduce it since it seems to be happening on the data path.

Perhaps we can move the encode/decode logic closer to the metadata abstraction to reuse the data structures to start.

Reply:
Create a Helper class, add non-static fields for the FW. Add encode/decode